neuraxle/pipeline.py

```diff
@@ -92,80 +98,114 @@ def fit(self, data_inputs, expected_outputs=None) -> 'Pipeline':
         :param expected_outputs: the expected data output to fit on
         :return: the pipeline itself
         """
-        new_self, _ = self.fit_transform_steps(data_inputs, expected_outputs)
+        ids = self.hasher.hash(self.hyperparams, data_inputs)
+        data_container = DataContainer(ids=ids, data_inputs=data_inputs, expected_outputs=expected_outputs)
+        new_self, _ = self._fit_transform_core(data_container)
 
         return new_self
 
-    def fit_transform_steps(self, data_inputs, expected_outputs):
+    def inverse_transform_processed_outputs(self, processed_outputs) -> Any:
+        """
+        After transforming all data inputs, and obtaining a prediction, we can inverse transform the processed outputs
+        :param processed_outputs: the forward transformed data input
+        :return: backward transformed processed outputs
+        """
+        for step_name, step in list(reversed(self.steps_as_tuple)):
+            processed_outputs = step.transform(processed_outputs)
+        return processed_outputs
+
+    def handle_fit_transform(self, data_container: DataContainer) -> ('BaseStep', Any):
+        """
+        Fit transform then rehash ids with hyperparams and transformed data inputs
+        :param data_container: data container to fit transform
+        :return: tuple(fitted pipeline, transformed data container)
+        """
+        new_self, data_container = self._fit_transform_core(data_container)
+
+        ids = self.hasher.rehash(data_container.ids, self.hyperparams, data_container.data_inputs)
+        data_container.set_ids(ids)
+
+        return new_self, data_container
+
+    def handle_transform(self, data_container: DataContainer) -> Any:
+        """
+        Transform then rehash ids with hyperparams and transformed data inputs
+        :param data_container: data container to transform
+        :return: tuple(fitted pipeline, transformed data container)
+        """
+        data_container = self._transform_core(data_container)
+
+        ids = self.hasher.rehash(data_container.ids, self.hyperparams, data_container.data_inputs)
+        data_container.set_ids(ids)
+
+        return data_container
+
+    def _fit_transform_core(self, data_container) -> ('Pipeline', DataContainer):
         """
         After loading the last checkpoint, fit transform each pipeline steps
-        :param data_inputs: the data input to fit on
-        :param expected_outputs: the expected data output to fit on
-        :return: the pipeline itself
+        :param data_container: the data container to fit transform on
+        :return: tuple(pipeline, data_container)
         """
-        steps_left_to_do, data_inputs = self.read_checkpoint(data_inputs)
+        steps_left_to_do, data_container = self._load_pipeline_checkpoint(data_container)
```

Comment:
We'll need to start thinking how (and if) we resume the steps here or something.